File: examples/puddy/ts/networking/session-manager.ts
// SessionManager — orchestrates agent sessions using assert/retract for state.
// All session state lives in the fact database. Only streaming accumulators
// (which need incremental appending) live in JS.

import { hold, claim, assert, retract } from "@jam/types";
import { type AgentEvent } from "../models/events";
import { isTerminalStatus } from "../models/session";
import { SandboxAgentClient, SandboxAgentError, type AgentInfo } from "./client";

let _nextMsgId = 0;
let _nextUserMsgId = 0;

export class SessionManager {
  private client: SandboxAgentClient;

  // Minimal JS-only state: streaming accumulators (can't incrementally append to a fact)
  private streamingText: Map<string, string> = new Map();
  private streamingThought: Map<string, string> = new Map();
  // Track session IDs and their current status for retract-before-assert pattern
  private sessionStatuses: Map<string, string> = new Map();
  // Track old plan entry counts so we can retract them
  private planEntryCounts: Map<string, number> = new Map();

  isConnected = false;
  connectionError?: string;
  pingMs?: number;
  agents: AgentInfo[] = [];

  constructor(client?: SandboxAgentClient) {
    this.client = client ?? new SandboxAgentClient();
  }

  get hostname(): string {
    return this.client.hostname;
  }

  get hasReadyAgent(): boolean {
    return this.agents.some((a) => a.installed && a.credentialsAvailable);
  }

  get preferredAgent(): AgentInfo | undefined {
    return this.agents.find((a) => a.installed && a.credentialsAvailable);
  }

  hasSession(id: string): boolean {
    return this.sessionStatuses.has(id);
  }

  get agentReadinessError(): string | undefined {
    if (!this.isConnected) return undefined;
    if (this.agents.length === 0) return "No agents found on server";
    const installed = this.agents.filter((a) => a.installed);
    if (installed.length === 0)
      return "No agents installed — run the sandbox-agent install command";
    const withCreds = installed.filter((a) => a.credentialsAvailable);
    if (withCreds.length === 0) {
      const names = installed.map((a) => a.id).join(", ");
      return `Agents installed (${names}) but no API credentials configured`;
    }
    return undefined;
  }

  // --- Connection ---

  async checkConnection(): Promise<void> {
    const start = Date.now();
    try {
      this.isConnected = await this.client.checkHealth();
      this.pingMs = Date.now() - start;
      this.connectionError = undefined;
      this.agents = await this.client.listAgents();
    } catch (err: any) {
      this.isConnected = false;
      this.pingMs = undefined;
      this.agents = [];
      this.connectionError = err.message ?? String(err);
    }
    this.syncConnectionState();
  }

  // --- Session Lifecycle ---

  createNewSession(initialPrompt?: string): string {
    const agent = this.preferredAgent?.id;
    if (!agent) throw SandboxAgentError.noReadyAgent();

    const sessionId = "s-" + Date.now();

    // Assert session facts directly into the database
    assert("session", sessionId, "agent", agent);
    assert("session", sessionId, "status", "starting");
    this.sessionStatuses.set(sessionId, "starting");

    // Async: connect to backend
    this.connectSession(sessionId, agent, initialPrompt);

    return sessionId;
  }

  private async connectSession(sessionId: string, agent: string, initialPrompt?: string): Promise<void> {
    try {
      await this.client.createSession(sessionId, agent);

      this.setStatus(sessionId, "active");

      if (initialPrompt) {
        this.sendMessage(sessionId, initialPrompt);
      }
    } catch (err: any) {
      this.setStatus(sessionId, "failed");
      assert("session", sessionId, "statusDetail", err.message ?? String(err));
    }
  }

  sendMessage(sessionId: string, message: string): void {
    // Add user message
    const msgId = `umsg-${_nextUserMsgId++}`;
    assert("message", sessionId, msgId, "user", "text", message);

    // Send prompt with SSE event capture — events arrive incrementally
    this.client.sendPromptWithEvents(
      sessionId,
      message,
      (event) => this.handleEvent(sessionId, event),
      () => {
        // POST completed — nothing to do, events already applied
      },
      (err) => {
        console.error("sendMessage error:", err.message ?? err);
        this.clearStreaming(sessionId);
      },
    );
  }

  private handleEvent(sessionId: string, event: AgentEvent): void {
    const p = event.payload;

    switch (p.type) {
      case "agentMessageChunk": {
        this.ensureActive(sessionId);
        const oldText = this.streamingText.get(sessionId);
        const newText = (oldText ?? "") + p.text;
        if (oldText) retract("session", sessionId, "streamingText", oldText);
        assert("session", sessionId, "streamingText", newText);
        this.streamingText.set(sessionId, newText);
        break;
      }

      case "agentThoughtChunk": {
        this.ensureActive(sessionId);
        const oldThought = this.streamingThought.get(sessionId);
        const newThought = (oldThought ?? "") + p.text;
        if (oldThought) retract("session", sessionId, "streamingThought", oldThought);
        assert("session", sessionId, "streamingThought", newThought);
        this.streamingThought.set(sessionId, newThought);
        break;
      }

      case "toolCall":
        this.finalizeStreaming(sessionId);
        assert("message", sessionId, p.data.toolCallId, "assistant", "toolUse", p.data.title);
        assert("message", sessionId, p.data.toolCallId, "toolStatus", p.data.status ?? "pending");
        assert("session", sessionId, "hasActiveTools", "true");
        break;

      case "toolCallUpdate": {
        const status = p.data.status ?? "in_progress";
        // Update tool status
        retract("message", sessionId, p.data.toolCallId, "toolStatus", p.data.status ?? "pending");
        assert("message", sessionId, p.data.toolCallId, "toolStatus", status);
        // Add result message for completed/failed
        if (status === "completed" || status === "failed") {
          assert("message", sessionId, p.data.toolCallId + "-result", "tool", "toolResult", status);
        }
        break;
      }

      case "usageUpdate":
        // Usage is transient — use hold so it auto-replaces
        hold(`usage-${sessionId}`, () => {
          if (p.data.costAmount != null) {
            claim("session", sessionId, "costAmount", p.data.costAmount);
            claim("session", sessionId, "costCurrency", p.data.costCurrency ?? "USD");
          }
          claim("session", sessionId, "contextSize", p.data.size);
          claim("session", sessionId, "contextUsed", p.data.used);
        });
        break;

      case "plan": {
        this.ensureActive(sessionId);
        // Retract old plan entries
        const oldCount = this.planEntryCounts.get(sessionId) ?? 0;
        for (let i = 0; i < oldCount; i++) {
          // We can't retract individual entries easily without knowing their content,
          // so use hold() for the plan which auto-retracts the previous set.
        }
        // Use hold for plan entries — they get replaced as a whole each time
        hold(`plan-${sessionId}`, () => {
          for (let i = 0; i < p.data.entries.length; i++) {
            const entry = p.data.entries[i];
            claim("plan", sessionId, `entry-${i}`, entry.content, entry.status, entry.priority);
          }
          claim("plan", sessionId, "count", String(p.data.entries.length));
        });
        this.planEntryCounts.set(sessionId, p.data.entries.length);
        break;
      }

      case "currentModeUpdate":
        assert("session", sessionId, "currentMode", p.modeId);
        // Add mode change message
        const modeId = `mode-${_nextMsgId++}`;
        assert("message", sessionId, modeId, "system", "modeChange", p.modeId);
        break;

      case "availableCommandsUpdate":
        // Use hold for commands — they get replaced as a whole
        hold(`commands-${sessionId}`, () => {
          for (const cmd of p.commands) {
            claim("command", sessionId, cmd.name, cmd.description);
          }
        });
        break;

      case "sessionInfoUpdate":
        if (p.title) {
          assert("session", sessionId, "title", p.title);
        }
        break;

      case "sessionEnd":
        this.finalizeStreaming(sessionId);
        this.setStatus(sessionId, "ended");
        assert("session", sessionId, "statusDetail", p.stopReason);
        // Clear active tools indicator
        retract("session", sessionId, "hasActiveTools", "true");
        break;

      case "unknown":
        break;
    }
  }

  async destroySession(id: string): Promise<void> {
    // Clean up JS-side state
    this.streamingText.delete(id);
    this.streamingThought.delete(id);
    this.sessionStatuses.delete(id);
    this.planEntryCounts.delete(id);

    // Clean up holds
    hold(`plan-${id}`, () => {});
    hold(`commands-${id}`, () => {});
    hold(`usage-${id}`, () => {});

    // Note: assert()ed facts for this session remain in the DB.
    // The UI just won't display them if the session is removed from the sidebar.
    // A full cleanup would require tracking all asserted facts per session.

    try {
      await this.client.destroySession(id);
    } catch {
      // best effort
    }
  }

  // --- Helpers ---

  private setStatus(sessionId: string, status: string): void {
    const old = this.sessionStatuses.get(sessionId);
    if (old) retract("session", sessionId, "status", old);
    assert("session", sessionId, "status", status);
    this.sessionStatuses.set(sessionId, status);
  }

  private ensureActive(sessionId: string): void {
    if (this.sessionStatuses.get(sessionId) === "starting") {
      this.setStatus(sessionId, "active");
    }
  }

  private finalizeStreaming(sessionId: string): void {
    // Finalize streaming text into a message
    const text = this.streamingText.get(sessionId);
    if (text) {
      retract("session", sessionId, "streamingText", text);
      assert("message", sessionId, `msg-${_nextMsgId++}`, "assistant", "text", text);
      this.streamingText.delete(sessionId);
    }

    // Finalize streaming thought into a message
    const thought = this.streamingThought.get(sessionId);
    if (thought) {
      retract("session", sessionId, "streamingThought", thought);
      assert("message", sessionId, `thought-${_nextMsgId++}`, "assistant", "thought", thought);
      this.streamingThought.delete(sessionId);
    }
  }

  private clearStreaming(sessionId: string): void {
    const text = this.streamingText.get(sessionId);
    if (text) {
      retract("session", sessionId, "streamingText", text);
      this.streamingText.delete(sessionId);
    }
    const thought = this.streamingThought.get(sessionId);
    if (thought) {
      retract("session", sessionId, "streamingThought", thought);
      this.streamingThought.delete(sessionId);
    }
  }

  // --- Connection State (uses hold since it's a single mutable value that replaces itself) ---

  private syncConnectionState(): void {
    hold("connection", () => {
      claim("connection", "status", this.isConnected ? "connected" : "disconnected");
      claim("connection", "hostname", this.hostname);
      if (this.connectionError) {
        claim("connection", "error", this.connectionError);
      }
    });
  }

  disconnectAll(): void {}
}

File: examples/puddy-native/PuddyApp/SessionManager.swift
import Foundation
import JamNative

/// Swift-side session manager — handles HTTP networking and SSE streaming
/// that can't run in JSC (no fetch API). Receives action callbacks from JS
/// via JamRuntime.onNativeAction (no polling).
final class SessionManager {
    let runtime: JamRuntime
    let baseURL: String
    private var sseTasks: [String: Task<Void, Never>] = [:]
    private var nextMsgId = 0

    init(runtime: JamRuntime, baseURL: String = "") {
        self.runtime = runtime
        self.baseURL = baseURL

        // Register the native action handler — JS calls callNative("action", {params})
        // which synchronously invokes this closure on the JS queue.
        runtime.onNativeAction = { [weak self] action, params in
            self?.handleNativeAction(action, params: params)
        }
    }

    /// Handle actions dispatched from JS via callNative().
    private func handleNativeAction(_ action: String, params: [String: Any]?) -> Any? {
        switch action {
        case "createSession":
            createSession()
            return nil

        case "sendMessage":
            let sessionId = params?["sessionId"] as? String ?? ""
            let text = params?["text"] as? String ?? ""
            if !sessionId.isEmpty && !text.isEmpty {
                sendMessage(sessionId, text)
            }
            return nil

        case "destroySession":
            let sessionId = params?["sessionId"] as? String ?? ""
            if !sessionId.isEmpty {
                Task { await destroySession(sessionId) }
            }
            return nil

        default:
            print("[Puddy] Unknown native action: \(action)")
            return nil
        }
    }

    // MARK: - Connection

    func checkConnection() {
        Task {
            do {
                let url = URL(string: "\(baseURL)/v1/health")!
                let (_, response) = try await URLSession.shared.data(from: url)
                let httpResponse = response as? HTTPURLResponse
                if httpResponse?.statusCode == 200 {
                    runtime.loadProgram(id: "connection-update", source: """
                        set("connection", "status", "connected");
                    """)
                    await fetchAgents()
                } else {
                    runtime.loadProgram(id: "connection-update", source: """
                        set("connection", "status", "disconnected");
                    """)
                }
            } catch {
                runtime.loadProgram(id: "connection-update", source: """
                    set("connection", "status", "disconnected");
                """)
            }
        }
    }

    private func fetchAgents() async {
        do {
            let url = URL(string: "\(baseURL)/v1/agents")!
            let (data, _) = try await URLSession.shared.data(from: url)
            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let agents = json["agents"] as? [[String: Any]],
               let first = agents.first(where: {
                   ($0["installed"] as? Bool) == true &&
                   ($0["credentialsAvailable"] as? Bool) == true
               }),
               let agentId = first["id"] as? String {
                runtime.loadProgram(id: "agent-update", source: """
                    set("connection", "agentId", "\(agentId)");
                """)
            }
        } catch {
            print("[Puddy] Failed to fetch agents: \(error)")
        }
    }

    // MARK: - Session Lifecycle

    func createSession() {
        let sessionId = "s-\(Int(Date().timeIntervalSince1970 * 1000))"

        runtime.loadProgram(id: "create-session-\(sessionId)", source: """
            var agentMatches = when(["connection", "agentId", $.id]);
            var agentId = agentMatches.length > 0 ? agentMatches[0].id : "unknown";
            assert("session", "\(sessionId)", "agent", agentId);
            assert("session", "\(sessionId)", "status", "starting");
            set("ui", "selectedSession", "\(sessionId)");
        """)

        Task { await connectSession(sessionId) }
    }

    private func connectSession(_ sessionId: String) async {
        let facts = runtime.getCurrentFacts()
        let agentId = extractFactValue(from: facts, entity: "connection", attr: "agentId") ?? "unknown"

        do {
            let url = URL(string: "\(baseURL)/v1/acp/\(sessionId)?agent=\(agentId)")!
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            let rpcBody: [String: Any] = [
                "jsonrpc": "2.0", "id": 1,
                "method": "session/new",
                "params": ["cwd": "/", "mcpServers": [Any]()]
            ]
            request.httpBody = try JSONSerialization.data(withJSONObject: rpcBody)

            let (_, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                setStatus(sessionId, "failed")
                return
            }

            setStatus(sessionId, "active")
            startEventStream(sessionId)
        } catch {
            setStatus(sessionId, "failed")
        }
    }

    func sendMessage(_ sessionId: String, _ text: String) {
        let msgId = "umsg-\(nextMsgId)"
        nextMsgId += 1
        let escaped = escapeJS(text)

        runtime.loadProgram(id: "send-msg-\(msgId)", source: """
            assert("message", "\(sessionId)", "\(msgId)", "user", "text", "\(escaped)");
            assert("session", "\(sessionId)", "thinking", "true");
        """)

        Task {
            let url = URL(string: "\(baseURL)/v1/acp/\(sessionId)")!
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            let rpcBody: [String: Any] = [
                "jsonrpc": "2.0", "id": 2,
                "method": "session/prompt",
                "params": ["prompt": [["type": "text", "text": text]]]
            ]
            request.httpBody = try? JSONSerialization.data(withJSONObject: rpcBody)

            do {
                let (_, response) = try await URLSession.shared.data(for: request)
                if (response as? HTTPURLResponse)?.statusCode != 200 {
                    print("[Puddy] sendMessage failed")
                }
            } catch {
                print("[Puddy] sendMessage error: \(error)")
            }
        }
    }

    func destroySession(_ id: String) async {
        sseTasks[id]?.cancel()
        sseTasks.removeValue(forKey: id)
        try? await URLSession.shared.data(for: URLRequest(url: URL(string: "\(baseURL)/v1/acp/\(id)")!))
    }

    // MARK: - SSE Event Stream

    private func startEventStream(_ sessionId: String) {
        sseTasks[sessionId]?.cancel()

        let task = Task {
            let url = URL(string: "\(baseURL)/v1/acp/\(sessionId)")!
            var request = URLRequest(url: url)
            request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
            request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")

            do {
                let (bytes, _) = try await URLSession.shared.bytes(for: request)
                for try await line in bytes.lines {
                    if Task.isCancelled { break }
                    guard line.hasPrefix("data: ") else { continue }
                    let data = String(line.dropFirst(6)).trimmingCharacters(in: .whitespaces)
                    if data.isEmpty { continue }
                    handleSSEData(sessionId, data)
                }
            } catch {
                if !Task.isCancelled {
                    print("[Puddy] SSE error for \(sessionId): \(error)")
                }
            }
        }
        sseTasks[sessionId] = task
    }

    private func handleSSEData(_ sessionId: String, _ data: String) {
        guard let jsonData = data.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { return }

        if json["error"] != nil { return }

        // Session end
        if json["id"] != nil, let result = json["result"] as? [String: Any],
           let stopReason = result["stopReason"] as? String {
            finalizeStreaming(sessionId)
            setStatus(sessionId, "ended")
            runtime.loadProgram(id: "session-end-\(sessionId)", source: """
                assert("session", "\(sessionId)", "statusDetail", "\(escapeJS(stopReason))");
                retract("session", "\(sessionId)", "hasActiveTools", "true");
            """)
            return
        }

        guard let params = json["params"] as? [String: Any],
              let update = params["update"] as? [String: Any],
              let sessionUpdate = update["sessionUpdate"] as? String else { return }

        let msgId = "msg-\(nextMsgId)"
        nextMsgId += 1

        switch sessionUpdate {
        case "agent_message_chunk":
            let text = (update["content"] as? [String: Any])?["text"] as? String ?? ""
            let escaped = escapeJS(text)
            runtime.loadProgram(id: "chunk-\(msgId)", source: """
                retract("session", "\(sessionId)", "thinking", "true");
                var old = when(["session", "\(sessionId)", "streamingText", $.t]);
                if (old.length > 0) retract("session", "\(sessionId)", "streamingText", old[0].t);
                var prev = old.length > 0 ? old[0].t : "";
                assert("session", "\(sessionId)", "streamingText", prev + "\(escaped)");
            """)

        case "agent_thought_chunk":
            let text = (update["content"] as? [String: Any])?["text"] as? String ?? ""
            let escaped = escapeJS(text)
            runtime.loadProgram(id: "thought-\(msgId)", source: """
                retract("session", "\(sessionId)", "thinking", "true");
                var old = when(["session", "\(sessionId)", "streamingThought", $.t]);
                if (old.length > 0) retract("session", "\(sessionId)", "streamingThought", old[0].t);
                var prev = old.length > 0 ? old[0].t : "";
                assert("session", "\(sessionId)", "streamingThought", prev + "\(escaped)");
            """)

        case "tool_call":
            let toolCallId = update["toolCallId"] as? String ?? msgId
            let title = escapeJS(update["title"] as? String ?? "Unknown tool")
            finalizeStreaming(sessionId)
            runtime.loadProgram(id: "tool-\(msgId)", source: """
                assert("message", "\(sessionId)", "\(toolCallId)", "assistant", "toolUse", "\(title)");
                assert("session", "\(sessionId)", "hasActiveTools", "true");
            """)

        case "tool_call_update":
            let toolCallId = update["toolCallId"] as? String ?? ""
            let status = update["status"] as? String ?? "in_progress"
            if status == "completed" || status == "failed" {
                runtime.loadProgram(id: "toolresult-\(msgId)", source: """
                    assert("message", "\(sessionId)", "\(toolCallId)-result", "tool", "toolResult", "\(status)");
                """)
            }

        case "plan":
            if let entries = update["entries"] as? [[String: Any]] {
                var source = "transaction(function() {\n"
                source += "  retract(\"plan\", \"\(sessionId)\", _, _, _, _);\n"
                for (i, entry) in entries.enumerated() {
                    let content = escapeJS(entry["content"] as? String ?? "")
                    let estatus = entry["status"] as? String ?? "pending"
                    let priority = entry["priority"] as? String ?? "medium"
                    source += "  assert(\"plan\", \"\(sessionId)\", \"entry-\(i)\", \"\(content)\", \"\(estatus)\", \"\(priority)\");\n"
                }
                source += "});"
                runtime.loadProgram(id: "plan-\(msgId)", source: source)
            }

        case "session_info_update":
            if let title = update["title"] as? String {
                runtime.loadProgram(id: "info-\(msgId)", source: """
                    set("session", "\(sessionId)", "title", "\(escapeJS(title))");
                """)
            }

        case "current_mode_update":
            let modeId = update["modeId"] as? String ?? update["currentModeId"] as? String ?? ""
            runtime.loadProgram(id: "mode-\(msgId)", source: """
                set("session", "\(sessionId)", "currentMode", "\(modeId)");
            """)

        default:
            break
        }
    }

    // MARK: - Helpers

    private func setStatus(_ sessionId: String, _ status: String) {
        runtime.loadProgram(id: "status-\(sessionId)-\(status)", source: """
            var old = when(["session", "\(sessionId)", "status", $.s]);
            if (old.length > 0) retract("session", "\(sessionId)", "status", old[0].s);
            assert("session", "\(sessionId)", "status", "\(status)");
        """)
    }

    private func finalizeStreaming(_ sessionId: String) {
        runtime.loadProgram(id: "finalize-\(sessionId)-\(nextMsgId)", source: """
            retract("session", "\(sessionId)", "thinking", "true");
            var streamText = when(["session", "\(sessionId)", "streamingText", $.t]);
            if (streamText.length > 0) {
                retract("session", "\(sessionId)", "streamingText", streamText[0].t);
                assert("message", "\(sessionId)", "msg-finalized-\(nextMsgId)", "assistant", "text", streamText[0].t);
            }
            var streamThought = when(["session", "\(sessionId)", "streamingThought", $.t]);
            if (streamThought.length > 0) {
                retract("session", "\(sessionId)", "streamingThought", streamThought[0].t);
                assert("message", "\(sessionId)", "thought-finalized-\(nextMsgId)", "assistant", "thought", streamThought[0].t);
            }
        """)
    }

    private func extractFactValue(from factsJson: String, entity: String, attr: String) -> String? {
        guard let data = factsJson.data(using: .utf8),
              let facts = try? JSONSerialization.jsonObject(with: data) as? [[Any]] else { return nil }
        for fact in facts {
            if fact.count >= 3, "\(fact[0])" == entity, "\(fact[1])" == attr {
                return "\(fact[2])"
            }
        }
        return nil
    }

    private func escapeJS(_ str: String) -> String {
        str.replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
            .replacingOccurrences(of: "\t", with: "\\t")
    }
}
